pkg/destroy/aws/aws.go

```diff
@@ -641,35 +646,110 @@ func deleteEC2Instance(ec2Client *ec2.EC2, iamClient *iam.IAM, id string, logger
 
 	for _, reservation := range response.Reservations {
 		for _, instance := range reservation.Instances {
-			// Skip 'terminated' instances since they take a while to really get cleaned up
-			if *instance.State.Name == "terminated" {
-				continue
-			}
-			if instance.IamInstanceProfile != nil {
-				parsed, err := arn.Parse(*instance.IamInstanceProfile.Arn)
-				if err != nil {
-					return errors.Wrap(err, "parse ARN for IAM instance profile")
-				}
-
-				err = deleteIAMInstanceProfile(iamClient, parsed, logger)
-				if err != nil {
-					return errors.Wrapf(err, "deleting %s", parsed.String())
-				}
-			}
-
-			_, err := ec2Client.TerminateInstances(&ec2.TerminateInstancesInput{
-				InstanceIds: []*string{instance.InstanceId},
-			})
+			err = deleteEC2InstanceByInstance(ec2Client, iamClient, instance, logger)
 			if err != nil {
 				return err
 			}
+		}
+	}
+	return nil
+}
+
+func deleteEC2InstanceByInstance(ec2Client *ec2.EC2, iamClient *iam.IAM, instance *ec2.Instance, logger logrus.FieldLogger) error {
+	// Skip 'terminated' instances since they take a while to really get cleaned up
+	if instance.State == nil || *instance.State.Name == "terminated" {
+		return nil
+	}
 
-			logger.Info("Deleted")
+	if instance.IamInstanceProfile != nil {
+		parsed, err := arn.Parse(*instance.IamInstanceProfile.Arn)
+		if err != nil {
+			return errors.Wrap(err, "parse ARN for IAM instance profile")
+		}
+
+		err = deleteIAMInstanceProfile(iamClient, parsed, logger)
+		if err != nil {
+			return errors.Wrapf(err, "deleting %s", parsed.String())
 		}
 	}
+
+	if *instance.State.Name == "shutting-down" {
```

Comment:
what corner-case is this trying to catch? there's already terminating check here https://github.com/openshift/installer/pull/2169/commits/50893d905487dca3da3d2963ceb4faf9a2441484#diff-ab65f0e2ba0237c5cc1429673c22cc08R660

Reply:
> what corner-case is this trying to catch? there's already terminating check here 50893d9#diff-ab65f0e2ba0237c5cc1429673c22cc08R660

While the main entry point for this package will flow in through `deleteEC2InstanceByInstance` with this PR, you can also get to this `deleteEC2InstanceByInstance` code via the `deleteARN` chain.  This guard makes sure we don't bother sending useless terminations to already-shutting-down instances regardless of how we enter this function.  So basically future-proofing against refactoring.  Or maybe against instances tagged into this cluster magically getting injected later in the teardown process?

Reply:
so why is after the `instance.IamInstanceProfile != nil` check?? and not with the state conditional on the top??

Reply:
> so why is after the `instance.IamInstanceProfile != nil` check?? and not with the state conditional on the top??

Good point.  Shifted to the top with a128b4350 -> 6af061afd.